WPKorg35/WPKorg35.cpp

```diff
@@ -0,0 +1,298 @@
+
+// Saturation is a scalar gain and 
+// can be used to drive the signal
+// into the non linear saturator
+CK_DLL_MFUN(wpkorg35_setSaturation);
+CK_DLL_MFUN(wpkorg35_getSaturation);
+
+// Enable non linear saturator
+CK_DLL_MFUN(wpkorg35_setNonLinearProcessing);
+CK_DLL_MFUN(wpkorg35_getNonLinearProcessing);
+
+// for Chugins extending UGen, this is mono synthesis function for 1 sample
+CK_DLL_TICK(wpkorg35_tick);
+
+// this is a special offset reserved for Chugin internal data
+t_CKINT wpkorg35_data_offset = 0;
+
+
+// class definition for internal Chugin data
+// (note: this isn't strictly necessary, but serves as example
+// of one recommended approach)
+class WPKorg35
+{
+public:
+    // constructor
+    WPKorg35( t_CKFLOAT fs)
+    {
+        korgFilter.prepareForPlay(fs);
+    }
+
+    // for Chugins extending UGen
+    SAMPLE tick( SAMPLE in )
+    {
+        // default: this passes whatever input is patched into Chugin
+        return korgFilter.doFilter(in);
+    }
+
+    float setResonance( t_CKFLOAT p )
+    {
+        korgFilter.m_dK = p;
+        korgFilter.updateFilters();
+        return p;
+    }
+
+    float getResonance() { return korgFilter.m_dK; }
+
+    float setCutoff( t_CKFLOAT p )
+    {
+        korgFilter.m_dFc = p;
+        korgFilter.updateFilters();
+        return p;
+    }
+
+    float getCutoff() { return korgFilter.m_dFc; }
+
+    float setSaturation( t_CKFLOAT p )
+    {
+        korgFilter.m_dSaturation = p;
+        return p;
+    }
+
+    float getSaturation() { return korgFilter.m_dSaturation; }
+    
+    bool setNonLinearProcessing( t_CKBOOL p )
+    {
+        korgFilter.m_uNonLinearProcessing = p;
+        return p;
+    }
+
+    bool getNonLinearProcessing() { return korgFilter.m_uNonLinearProcessing; }
+    
+private:
+    // instance data
+    CKorgThreeFiveLPF korgFilter;
+};
+
+
+// query function: chuck calls this when loading the Chugin
+// NOTE: developer will need to modify this function to
+// add additional functions to this Chugin
+CK_DLL_QUERY( WPKorg35 )
+{
+    // hmm, don't change this...
+    QUERY->setname(QUERY, "WPKorg35");
+    
+    // begin the class definition
+    // can change the second argument to extend a different ChucK class
+    QUERY->begin_class(QUERY, "WPKorg35", "UGen");
+
+    // register the constructor (probably no need to change)
+    QUERY->add_ctor(QUERY, wpkorg35_ctor);
+    // register the destructor (probably no need to change)
+    QUERY->add_dtor(QUERY, wpkorg35_dtor);
+    
+    // for UGen's only: add tick function
+    QUERY->add_ugen_func(QUERY, wpkorg35_tick, NULL, 1, 1);
+    
+    // NOTE: if this is to be a UGen with more than 1 channel, 
+    // e.g., a multichannel UGen -- will need to use add_ugen_funcf()
+    // and declare a tickf function using CK_DLL_TICKF
+
+    // example of adding setter method
+    QUERY->add_mfun(QUERY, wpkorg35_setResonance, "float", "resonance");
+    // example of adding argument to the above method
```

Comment:
Cool, I'd add some documentation to the general README too.

On Sat, Sep 5, 2015 at 12:39 PM, owenvallis <notifications@github.com>
wrote:

> In WPKorg35/WPKorg35.cpp
> <https://github.com/ccrma/chugins/pull/16#discussion_r38814144>:
>
> > +
> > +    // register the constructor (probably no need to change)
> > +    QUERY->add_ctor(QUERY, wpkorg35_ctor);
> > +    // register the destructor (probably no need to change)
> > +    QUERY->add_dtor(QUERY, wpkorg35_dtor);
> > +
> > +    // for UGen's only: add tick function
> > +    QUERY->add_ugen_func(QUERY, wpkorg35_tick, NULL, 1, 1);
> > +
> > +    // NOTE: if this is to be a UGen with more than 1 channel,
> > +    // e.g., a multichannel UGen -- will need to use add_ugen_funcf()
> > +    // and declare a tickf function using CK_DLL_TICKF
> > +
> > +    // example of adding setter method
> > +    QUERY->add_mfun(QUERY, wpkorg35_setResonance, "float", "resonance");
> > +    // example of adding argument to the above method
>
> I added descriptions at the top of the class. I'll move them to use the
> doc_func
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ccrma/chugins/pull/16/files#r38814144>.
>